language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -242,34 +259,34 @@ computeCek con t@(Var _ varName)   = do
--- Instantiate all the free variable of the resulting term in case there are any.
+-- Instantiate all the free variables of the resulting term in case there are any.
 returnCek [] res = getVarEnv <&> \varEnv -> void $ dischargeVarEnv varEnv res
-returnCek (FrameTyInstArg ty : con) fun = instantiateEvaluate con ty fun
-returnCek (FrameApplyArg argVarEnv arg : con) fun = do
+returnCek (FrameTyInstArg ty : ctx) fun =
+    case fun of
+      TyAbs _ _ _ body -> computeCek ctx body  -- NB: we're just ignoring the type here
+      _                ->  throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just (void fun)
+returnCek (FrameApplyArg argVarEnv arg : ctx) fun = do
     funVarEnv <- getVarEnv
-    withVarEnv argVarEnv $ computeCek (FrameApplyFun funVarEnv fun : con) arg
-returnCek (FrameApplyFun funVarEnv fun : con) arg = do
-    argVarEnv <- getVarEnv
-    applyEvaluate funVarEnv argVarEnv con fun arg
-returnCek (FrameIWrap ann pat arg : con) val =
-    returnCek con $ IWrap ann pat arg val
-returnCek (FrameUnwrap : con) dat = case dat of
-    IWrap _ _ _ term -> returnCek con term
+    withVarEnv argVarEnv $ computeCek (FrameApplyFun funVarEnv fun : ctx) arg
+returnCek (FrameApplyFun funVarEnv fun : ctx) arg =
+    case fun of
+      LamAbs _ name _ body -> do
+               argVarEnv <- getVarEnv
+               withVarEnv (extendVarEnv name arg argVarEnv funVarEnv) $ computeCek ctx body
+      _ ->  throwingWithCause _MachineError NonPrimitiveApplicationMachineError $ Just (Apply () (void fun) (void arg))
+      -- FIXME: ^ error message looks like eg (1 2) instead of [(con 1) (con 2)]
+returnCek (FrameIWrap ann pat arg : ctx) val =
+    returnCek ctx $ IWrap ann pat arg val
+returnCek (FrameUnwrap : ctx) dat = case dat of
+    IWrap _ _ _ term -> returnCek ctx term
     term             ->
         throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just (void term)
-
--- | Instantiate a term with a type and proceed.
--- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
--- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
--- apply the term to the type via 'TyInst'.
-instantiateEvaluate
-    :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => Context uni -> Type TyName uni ExMemory -> WithMemory Term uni -> CekM uni (Plain Term uni)
-instantiateEvaluate con _ (TyAbs _ _ _ body) = computeCek con body
-instantiateEvaluate con ty fun
-    | isJust $ termAsPrimIterApp fun = returnCek con $ TyInst (memoryUsage () <> memoryUsage fun <> memoryUsage ty) fun ty
-    | otherwise                      =
-        throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just (void fun)
+returnCek (FrameApplyBuiltin env bn tys closures terms : ctx) value = do
+    varEnv <- getVarEnv
+    let closures' = Closure varEnv value : closures -- Save the closure for the argument we've just computed.
```

Comment:
If we go with CPS, then we can store a continuation instead of a list of closures, because that is how the list of closures is going to be used down the control flow anyway.

Reply:
I don't think we even need CPS.  I think we can do it all with the frames (this is what the comment at the start of the definition of `getScopedArgs` was alluding to).  The frame for builtin application has to save the environment, `env0`, say that the application's being evaluated in because it needs it for each of that arguments.  When you compute an argument down to a value the `return` phase creates a closure which gets saved in a new frame and then it goes on to deal with the next argument.  I think that when you're returning a newly-computed argument value you can look at it to see if it's a constant or not, and if it's not constant, bind it to a variable and then extend `env0` with the new binding and use that to compute the next argument.  Now maybe that's not quite right, since the next argument is being computed in a larger environment than it should be.  I don't think that matters, because we've used new variables, but even if it does matter we can still accumulate the extended environment in the frame and at the end use that to carry on with the computation.  I didn't want to do that right away because while I was implementing this it was less complicated to deal with the arguments separately (and test that that method worked), but I think it should work.